exchanges/coinbasepro/coinbasepro_wrapper.go

```diff
@@ -596,16 +809,44 @@ func (c *CoinbasePro) WithdrawCryptocurrencyFunds(ctx context.Context, withdrawR
 	if err := withdrawRequest.Validate(); err != nil {
 		return nil, err
 	}
-	resp, err := c.WithdrawCrypto(ctx,
-		withdrawRequest.Amount,
-		withdrawRequest.Currency.String(),
-		withdrawRequest.Crypto.Address)
+	if withdrawRequest.WalletID == "" {
+		return nil, errWalletIDEmpty
+	}
+
+	creds, err := c.GetCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	message := fmt.Sprintf("%+v", withdrawRequest)
```

Comment:
👀

Reply:
Need to create an idempotency token; if we send the same one twice, the second transaction will be cancelled. I tried to generate it in a way that'd minimize the risk of collisions, without getting too big. Theoretically, we could create it in whichever way we wanted.

Reply:
Might be able to just use a unix nano timestamp?

Reply:
True, I can just use something quicker like that.
```go
t := time.Now().UnixNano()
u := math.Float64bits(withdrawRequest.Amount)
i := t ^ int64(u)
message := strconv.FormatInt(i, 10)
```
The intervening two lines cause no perceptible difference in speed; they're both really fast operations with data that we're already touching, and should hopefully deal with the chance of a freak accident collision.

Reply:
This is cool and I will use this somewhere 🔥 much better than what was happening before; 

If the amounts are exactly the same and by chance n * diff routines would retrieve the same unix timestamp it would yield the same deterministic result. 

I think below at least you can add a floor so you can guarantee what has been done in the past won't ever be used instead of pseudorandom numbers which could be hit in the future. 

```go
message := strconv.FormatInt(time.Now().UnixNano(), 10)
```

This is such an edge case for this function though, so you do you man. 🚀

Reply:
Yeah, still fails at same time & same amount, but two withdrawals with same float-precise amounts & nanosecond-precise times is a bit of an edge of an edge. There's not much more good entropy left ; least expensive thing left is the target walletID, which is still pretty prone to being static.

Even the old method of hashing the entire withdrawal request with the secret key wouldn't have added much personal anti-collision protection than this, I think. Exchange, currency, request type, description, and secret key can all be pretty static.